File: Breadth-First Search/Topological sort/LET_no785_Is Graph Bipartite?.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colors = [0] * n
        
        for i in range(n):
            if colors[i] != 0:
                continue
            queue = collections.deque([i])
            colors[i] = 1
            
            while queue:
                cur = queue.popleft()
                for nxt in graph[cur]:
                    if colors[nxt] == 0:
                        colors[nxt] = -colors[cur]
                        queue.append(nxt)
                    elif colors[nxt] != -colors[cur]:
                        return False
        
        return True

File: Breadth-First Search/Topological sort/LET_no785_Is Graph Bipartite?.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> nodes(n, 0);

        for (int i = 0; i < n; i++) {
            if (nodes[i] != 0)
                continue;

            nodes[i] = 1;
            queue<int> q;
            q.push(i);

            while (!q.empty()) {
                int u = q.front();
                q.pop();

                for (int v : graph[u]) {
                    if (nodes[v] == 0) {
                        nodes[v] = -nodes[u];
                        q.push(v);
                    }
                    else if (nodes[v] != -nodes[u])
                        return false;
                }
            }
        }

        return true;
    }
};

// A graph is bipartite if and only if the two ends of each edge have different colors.